### PR TITLE
Damien/lifecycle

### PR DESCRIPTION
**Pull request by dkoh555**
Successful implementation of Lifecycle nodes for Omnid startup; need to tidy up code and documentation

### CODE REVIEW

File: src/lifecycle_point_stamped_incr.cpp [hunk 1]
```diff
@@ -0,0 +1,955 @@
...
+float curr_y_max = 1.0;
+float curr_z_max = 1.0;
+static float x_max;
+static float y_max;
+static float z_max;
+static float alt_x_max;
+static float alt_y_max;
+static float alt_z_max;
+static float boundary_radius;
+float lin_rate_chg_fac;
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: static

File: src/lifecycle_point_stamped_incr.cpp [hunk 2]
```diff
@@ -0,0 +1,955 @@
...
+
+        /// @brief Default constructor. Initializes index to UNUSED_INDEX and type to InputType::None
+        MovementInput() : index(UNUSED_INDEX), type(InputType::None) {}
+        /// @brief Constructor with provided index and type parameters.
+        /// @param index_no The index of the joy message array.
+        /// @param input_type The type of input.
+        MovementInput(int index_no, InputType input_type) : index(index_no), type(input_type) {}
+};
+
+// /// Additional parameter helper functions developed my NU MSR
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: No commented-out code

File: src/lifecycle_point_stamped_incr.cpp [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+            declare_parameter("y_axis_inc", "UNUSED");
+            const std::string y_inc_assignment = get_parameter("y_axis_inc").as_string();
+            declare_parameter("y_axis_dec", "UNUSED");
+            const std::string y_dec_assignment = get_parameter("y_axis_dec").as_string();
+            declare_parameter("z_axis_inc", "UNUSED");
+            const std::string z_inc_assignment = get_parameter("z_axis_inc").as_string();
+            declare_parameter("z_axis_dec", "UNUSED");
+            const std::string z_dec_assignment = get_parameter("z_axis_dec").as_string();
+
+            // const std::string enable_assignment = rosnu::declare_and_get_param<std::string>("enable_control", "UNUSED", *this, "Button assigned to enable control inputs");
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: No commented out code

File: src/lifecycle_point_stamped_incr.cpp [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+            declare_parameter("alt_z_max", 0.);
+            alt_z_max = get_parameter("alt_z_max").as_double();
+            declare_parameter("x_flip", false);
+            x_flip = get_parameter("x_flip").as_bool();
+            declare_parameter("y_flip", false);
+            y_flip = get_parameter("y_flip").as_bool();
+            declare_parameter("z_flip", false);
+            z_flip = get_parameter("z_flip").as_bool();
+
+            // RCLCPP_INFO(rclcpp::get_logger("lifecycle_point_stamped_incr"), "TESTING PARAMETERS");
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: No commented out code

File: src/lifecycle_point_stamped_incr.cpp [hunk 5]
```diff
@@ -0,0 +1,955 @@
...
+
+            // x_max = rosnu::declare_and_get_param<float>("x_max", 1.0f, *this, "The maximum output value along that axis of movement");
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: No Commented out code

File: src/lifecycle_point_stamped_incr.cpp [hunk 6]
```diff
@@ -0,0 +1,955 @@
...
+            }
+
+            declare_parameter("x_offset", 0.);
+            x_offset = get_parameter("x_offset").as_double();
+            declare_parameter("y_offset", 0.);
+            y_offset = get_parameter("y_offset").as_double();
+            declare_parameter("z_offset", 0.);
+            z_offset = get_parameter("z_offset").as_double();
+
+            // x_offset = rosnu::declare_and_get_param<float>("x_offset", 0.0f, *this, "The offset for the message's zero value");
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: No commented out code

File: src/lifecycle_point_stamped_incr.cpp [hunk 7]
```diff
@@ -0,0 +1,955 @@
...
+            // RCLCPP_INFO(rclcpp::get_logger("lifecycle_point_stamped_incr"), "TEST JOY");
+            latest_joy_state = *joy_state;
+            fresh_joy_state = true;
+        }
+        rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_sub;
+};
+
+int main(int argc, char * argv[])
+{
+    // rclcpp::init(argc, argv);
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: No commented out code

File: src/lifecycle_point_stamped_incr.cpp [hunk 8]
```diff
@@ -0,0 +1,955 @@
...
+        case InputType::None:
+            break;
+        case InputType::Axis:
+            new_command.point.z = new_command.point.z + (curr_z_max/rate_of_change_denom) * lin_rate_chg_fac * (latest_joy_state.axes.at(input.index) > 0 ? latest_joy_state.axes.at(input.index) : 0) * pow(-1, z_flip);
+            break;
+        case InputType::Trigger:
+            new_command.point.z = new_command.point.z + (curr_z_max/rate_of_change_denom) * lin_rate_chg_fac * (0.5 - (latest_joy_state.axes.at(input.index)/2.0)) * pow(-1, z_flip);
+            break;
+        case InputType::Button:
+            new_command.point.z = new_command.point.z + (curr_z_max/rate_of_change_denom) * lin_rate_chg_fac * latest_joy_state.buttons.at(input.index) * pow(-1, z_flip);
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp:745]: Need default case, can throw std::logic_error if can't ever get there.

File: src/lifecycle_point_stamped_incr.cpp [hunk 9]
```diff
@@ -0,0 +1,955 @@
...
+            break;
+        case InputType::Axis:
+            new_command.point.z = new_command.point.z + (curr_z_max/rate_of_change_denom) * lin_rate_chg_fac * (latest_joy_state.axes.at(input.index) > 0 ? latest_joy_state.axes.at(input.index) : 0) * pow(-1, z_flip);
+            break;
+        case InputType::Trigger:
+            new_command.point.z = new_command.point.z - (curr_z_max/rate_of_change_denom) * lin_rate_chg_fac * (0.5 - (latest_joy_state.axes.at(input.index)/2.0)) * pow(-1, z_flip);
+            break;
+        case InputType::Button:
+            new_command.point.z = new_command.point.z - (curr_z_max/rate_of_change_denom) * lin_rate_chg_fac * latest_joy_state.buttons.at(input.index) * pow(-1, z_flip);
+            break;
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp:779]: default case

File: src/lifecycle_point_stamped_incr.cpp [hunk 10]
```diff
@@ -0,0 +1,955 @@
...
+    {
+        new_command.point.z = -curr_z_max;
+    }
+
+    return new_command;
+}
+
+static geometry_msgs::msg::PointStamped z_axis_dec(const MovementInput input, geometry_msgs::msg::PointStamped temp_command)
+{
+    geometry_msgs::msg::PointStamped new_command = temp_command;
```
Comment (m-elwin) [inline src/lifecycle_point_stamped_incr.cpp]: auto

File: src/lifecycle_twist_mirror.cpp [hunk 11]
```diff
@@ -0,0 +1,1528 @@
...
+    exe.add_node(lc_node->get_node_base_interface());
+
+    exe.spin();
+
+    rclcpp::shutdown();
+
+    return 0;
+}
+
+// int main(int argc, char * argv[])
```
Comment (m-elwin) [inline src/lifecycle_twist_mirror.cpp:722]: No commented out code

**Review by m-elwin — CHANGES_REQUESTED** [review]
This is a cursory review, but we will likely merge it and then have other improvements as separate issues.  However, the commented out code should be removed as it then will be confusing whether it was needed/temporarily commented out or what.

File: src/lifecycle_point_stamped_incr.cpp (hunk 1 above)
Comment (dkoh555) [reply src/lifecycle_point_stamped_incr.cpp]: I keep forgetting why static is useful :p but all global variables are now static

File: src/lifecycle_point_stamped_incr.cpp (hunk 2 above)
Comment (dkoh555) [reply src/lifecycle_point_stamped_incr.cpp]: Left a few commented out debugging log statements, but the rest is all gone

File: src/lifecycle_point_stamped_incr.cpp (hunk 4 above)
Comment (dkoh555) [reply src/lifecycle_point_stamped_incr.cpp]: Okay I removed all the commented out debugging statements

File: src/lifecycle_point_stamped_incr.cpp (hunk 10 above)
Comment (dkoh555) [reply src/lifecycle_point_stamped_incr.cpp]: Done

**Pull request merged**